File: 10-edge/02-continuous-batching-and-paged-kv/python/cancel_propagation.py
"""
Layer 10 - Topic 2: does hanging up actually free the KV blocks? (Python)

What this demonstrates
    One program, both halves of the failure. It stands up a stub model
    server that streams 40 tokens at 100ms each while watching for its
    caller to go away, puts a FastAPI gateway in front of it, then fires a
    client that hangs up after 500ms -- twice, at two gateway handlers:

      /naive      buffers the upstream response before returning anything.
                  It is the handler almost everyone writes first, and it
                  never learns the client left.
      /cancelling streams, polls `await request.is_disconnected()`, and
                  leaves the `async with client.stream(...)` block on
                  disconnect, which closes the upstream socket.

    The stub server reports what it saw. That report is the experiment:
    under load, a request whose client is gone but whose decode continues
    is holding KV blocks the scheduler could be giving to someone real.

What to look for
    - /naive: the upstream runs to completion, ~4s, roughly 3.5s of it
      after the client stopped listening.
    - /cancelling: the upstream sees EOF within a few tokens of the hang-up
      and stops.
    - The mechanism is not "FastAPI cancels for you". Starlette does cancel
      the handler task on http.disconnect, but that only surfaces at the
      next `await` -- so a handler that is inside a blocking call, or that
      awaits nothing until the upstream finishes, learns nothing. The
      polling in /cancelling is what makes the behaviour the same either
      way.

Dependencies (declared in requirements.txt): fastapi, uvicorn, httpx.
Runs with no arguments, binds only to 127.0.0.1 on ephemeral ports:

    python3 python/cancel_propagation.py
"""

from __future__ import annotations

import asyncio
import contextlib
import socket
import time

import httpx
import uvicorn
from fastapi import FastAPI, Request
from fastapi.responses import Response, StreamingResponse

TOKENS = 40
TOKEN_INTERVAL = 0.1          # 4.0s of "decode" in total
CLIENT_HANGS_UP_AFTER = 0.5   # the client gives up here

ledger: list[dict] = []


def free_port() -> int:
    with socket.socket() as s:
        s.bind(("127.0.0.1", 0))
        return s.getsockname()[1]


UPSTREAM_PORT = free_port()
GATEWAY_PORT = free_port()
UPSTREAM = f"http://127.0.0.1:{UPSTREAM_PORT}"


# --------------------------------------------------------------------------
# The stub model server. Streams tokens while concurrently watching its own
# socket for EOF, which is how a real engine learns that the consumer of a
# generation has gone away.
# --------------------------------------------------------------------------
async def upstream_handler(reader: asyncio.StreamReader,
                           writer: asyncio.StreamWriter) -> None:
    while await reader.readline() not in (b"\r\n", b"", b"\n"):
        pass  # drain request headers; the body is irrelevant to the point

    writer.write(b"HTTP/1.1 200 OK\r\nContent-Type: text/event-stream\r\n"
                 b"Transfer-Encoding: chunked\r\n\r\n")
    await writer.drain()

    peer_gone = asyncio.Event()

    async def watch_for_eof() -> None:
        # A read that returns b"" means the peer closed its end. This is the
        # only signal the engine gets, and it only arrives if somebody
        # upstream actually closes the socket.
        if await reader.read(1) == b"":
            peer_gone.set()

    watcher = asyncio.create_task(watch_for_eof())
    started = time.perf_counter()
    sent = 0
    try:
        for i in range(TOKENS):
            if peer_gone.is_set():
                break
            body = f"data: token {i}\n\n".encode()
            writer.write(b"%x\r\n" % len(body) + body + b"\r\n")
            try:
                await writer.drain()
            except (ConnectionResetError, BrokenPipeError):
                peer_gone.set()
                break
            sent = i + 1
            await asyncio.sleep(TOKEN_INTERVAL)
        writer.write(b"0\r\n\r\n")
        await writer.drain()
    except (ConnectionResetError, BrokenPipeError):
        peer_gone.set()
    finally:
        watcher.cancel()
        ledger.append({
            "aborted": peer_gone.is_set(),
            "tokens": sent,
            "seconds": time.perf_counter() - started,
        })
        writer.close()


# --------------------------------------------------------------------------
# The gateway: the same upstream call written two ways.
# --------------------------------------------------------------------------
app = FastAPI()

# One long-lived client, as a real service would have. Note that this alone
# is enough to defeat cancellation in /naive: nothing closes it, so nothing
# closes the upstream socket.
shared_client = httpx.AsyncClient(timeout=60.0)


@app.post("/naive")
async def naive() -> Response:
    """Await the whole upstream response, then return it. The shape almost
    everyone writes first, and it is not obviously wrong on the page.

    Starlette does not cancel a plain (non-streaming) handler when the
    client disconnects -- the http.disconnect message simply sits in the
    receive queue while this coroutine runs to completion. So the upstream
    generation runs to its end, holding KV blocks, for a response that is
    thrown away at the last line of this function."""
    r = await shared_client.post(f"{UPSTREAM}/completions", json={})
    return Response(content=r.content, media_type="text/event-stream")


@app.post("/cancelling")
async def cancelling(request: Request) -> StreamingResponse:
    """Stream, and hold the upstream response inside a scope that ends when
    this generator does.

    Two mechanisms are at work and it is worth knowing which is load-
    bearing. The explicit one is the `is_disconnected()` poll. The implicit
    one -- the one that actually fires first here -- is that Starlette
    throws CancelledError into an in-progress StreamingResponse generator
    when the client goes away, which unwinds `async with client.stream(...)`,
    which closes the upstream socket. Structure does the work; the poll is
    the backstop for the case where the upstream produces nothing for a long
    time and the generator is parked on a chunk that never arrives."""
    async def body():
        async with shared_client.stream("POST", f"{UPSTREAM}/completions",
                                        json={}) as upstream:
            async for chunk in upstream.aiter_bytes():
                if await request.is_disconnected():
                    return
                yield chunk
    return StreamingResponse(body(), media_type="text/event-stream")


async def hang_up_on(path: str) -> None:
    """A client that starts reading and then hangs up.

    Deliberately a raw socket rather than an HTTP client with a timeout.
    A timeout in httpx (and in most client libraries) raises in your code
    without necessarily closing the TCP connection promptly -- so the server
    sees nothing, and the experiment silently measures the wrong thing. A
    browser tab closing, or a proxy giving up, closes the socket. That close
    is the only signal the server ever gets, so it is the one to send."""
    reader, writer = await asyncio.open_connection("127.0.0.1", GATEWAY_PORT)
    writer.write(f"POST {path} HTTP/1.1\r\nHost: localhost\r\n"
                 f"Content-Type: application/json\r\nContent-Length: 2\r\n"
                 f"\r\n{{}}".encode())
    await writer.drain()
    with contextlib.suppress(asyncio.TimeoutError):
        await asyncio.wait_for(reader.read(), CLIENT_HANGS_UP_AFTER)
    writer.close()
    with contextlib.suppress(Exception):
        await writer.wait_closed()


async def main() -> None:
    upstream_server = await asyncio.start_server(upstream_handler, "127.0.0.1",
                                                 UPSTREAM_PORT)
    config = uvicorn.Config(app, host="127.0.0.1", port=GATEWAY_PORT,
                            log_level="error")
    gateway = uvicorn.Server(config)
    gateway_task = asyncio.create_task(gateway.serve())
    while not gateway.started:
        await asyncio.sleep(0.02)

    print("Python / FastAPI - cancellation on client disconnect")
    print(f"  upstream streams {TOKENS} tokens x {TOKEN_INTERVAL * 1000:.0f}ms "
          f"= {TOKENS * TOKEN_INTERVAL:.1f}s of decode")
    print(f"  client hangs up after {CLIENT_HANGS_UP_AFTER:.1f}s\n")
    print(f"  {'handler':<14} {'upstream saw':<16} {'tokens decoded':>14} "
          f"{'upstream ran':>13} {'wasted':>8}")
    print("  " + "-" * 70)

    for path in ("/naive", "/cancelling"):
        ledger.clear()
        await hang_up_on(path)
        # Give the stub a moment to finish or notice, whichever it does.
        deadline = time.perf_counter() + TOKENS * TOKEN_INTERVAL + 1.0
        while not ledger and time.perf_counter() < deadline:
            await asyncio.sleep(0.05)
        entry = ledger[0] if ledger else {"aborted": False, "tokens": -1,
                                          "seconds": float("nan")}
        wasted = max(0.0, entry["seconds"] - CLIENT_HANGS_UP_AFTER)
        print(f"  {path:<14} {'cancelled' if entry['aborted'] else 'nothing':<16} "
              f"{entry['tokens']:>14} {entry['seconds']:>12.2f}s "
              f"{wasted:>7.2f}s")

    print()
    print("  'wasted' is decode time spent on a response nobody read. On a")
    print("  loaded server that is not just wasted compute: those KV blocks")
    print("  stayed allocated the whole time, so the scheduler could not")
    print("  admit somebody who was still waiting.")

    gateway.should_exit = True
    await gateway_task
    upstream_server.close()
    await upstream_server.wait_closed()


if __name__ == "__main__":
    asyncio.run(main())

File: 10-edge/02-continuous-batching-and-paged-kv/nodejs/cancel_propagation.js
// Layer 10 - Topic 2: does hanging up actually free the KV blocks? (Node.js)
//
// What this demonstrates
//     The same experiment as python/cancel_propagation.py, in the runtime
//     where the answer depends on an AbortController you remembered to
//     wire up. A stub model server streams 40 tokens at 100ms each while
//     watching for its caller to leave. A gateway sits in front with two
//     handlers, and a client hangs up after 500ms against each:
//
//       /naive       `await fetch(upstream)` then `await r.arrayBuffer()`.
//                    No signal, so nothing connects the downstream socket
//                    closing to the upstream request. Generation runs to
//                    completion for a response that is discarded.
//       /cancelling  an AbortController aborted from `req.on('close')`,
//                    passed to fetch as `signal`. The abort tears down the
//                    upstream socket, and the engine sees EOF.
//
// What to look for
//     - /naive: upstream decodes all 40 tokens, ~3.5s of it after the
//       client stopped listening -- KV blocks held the whole time.
//     - /cancelling: upstream sees EOF within a token or two of the hang-up.
//     - The Node-specific caveat, which this file cannot show you but which
//       decides your real cancellation latency: `req.on('close')` is an
//       event-loop callback. Any synchronous CPU-bound stretch on the main
//       thread delays it along with everything else, so cancellation
//       latency here is bounded below by your worst blocking call. That is
//       Layer 1's lesson, arriving with a bill attached.
//
// No dependencies. Runs with no arguments, binds 127.0.0.1 only:
//     node nodejs/cancel_propagation.js

'use strict';

const http = require('node:http');
const net = require('node:net');

const TOKENS = 40;
const TOKEN_INTERVAL_MS = 100;        // 4.0s of "decode" in total
const CLIENT_HANGS_UP_AFTER_MS = 500;

const ledger = [];

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// ---------------------------------------------------------------------------
// The stub model server. Streams tokens; notices when its caller goes away.
// ---------------------------------------------------------------------------
const upstream = http.createServer(async (req, res) => {
  const started = process.hrtime.bigint();
  let sent = 0;
  let peerGone = false;

  // Fires when the connection closes, whether or not the response finished.
  res.on('close', () => {
    if (!res.writableFinished) peerGone = true;
  });

  res.writeHead(200, { 'Content-Type': 'text/event-stream' });
  for (let i = 0; i < TOKENS; i += 1) {
    if (peerGone) break;
    if (!res.write(`data: token ${i}\n\n`)) {
      // Backpressure, not disconnection -- but a write that throws EPIPE
      // lands in the 'error' handler below and means the same thing.
    }
    sent = i + 1;
    await sleep(TOKEN_INTERVAL_MS);
  }
  if (!peerGone) res.end();

  ledger.push({
    aborted: peerGone,
    tokens: sent,
    seconds: Number(process.hrtime.bigint() - started) / 1e9,
  });
});
upstream.on('clientError', () => {});

// ---------------------------------------------------------------------------
// The gateway: the same upstream call written two ways.
// ---------------------------------------------------------------------------
const gateway = http.createServer(async (req, res) => {
  const url = `http://127.0.0.1:${upstream.address().port}/completions`;

  if (req.url === '/naive') {
    // Buffer the whole response, then reply. Nothing in this shape can
    // observe the client, and no error is raised when it leaves: writing
    // to a closed response is a silent no-op in Node.
    try {
      const r = await fetch(url, { method: 'POST' });
      const body = await r.arrayBuffer();
      res.writeHead(200, { 'Content-Type': 'text/event-stream' });
      res.end(Buffer.from(body));
    } catch {
      res.destroy();
    }
    return;
  }

  // The fix, and it is four lines: one controller, one listener, one
  // signal, one guarded pipe.
  const ac = new AbortController();
  req.on('close', () => ac.abort());
  try {
    const r = await fetch(url, { method: 'POST', signal: ac.signal });
    res.writeHead(200, { 'Content-Type': 'text/event-stream' });
    for await (const chunk of r.body) {
      if (ac.signal.aborted) break;
      res.write(chunk);
    }
    res.end();
  } catch (err) {
    if (err.name !== 'AbortError') res.destroy();
  }
});

// ---------------------------------------------------------------------------
// The client that hangs up. A raw socket on purpose: a timeout in an HTTP
// client library raises in your code without necessarily closing the TCP
// connection, so the server sees nothing and the experiment measures the
// wrong thing. Closing the socket is the only signal a server ever gets.
// ---------------------------------------------------------------------------
function hangUpOn(path) {
  return new Promise((resolve) => {
    const sock = net.connect(gateway.address().port, '127.0.0.1', () => {
      sock.write(
        `POST ${path} HTTP/1.1\r\nHost: localhost\r\nContent-Length: 0\r\n\r\n`,
      );
      setTimeout(() => {
        sock.destroy();
        resolve();
      }, CLIENT_HANGS_UP_AFTER_MS);
    });
    sock.on('error', () => resolve());
  });
}

async function main() {
  await new Promise((r) => upstream.listen(0, '127.0.0.1', r));
  await new Promise((r) => gateway.listen(0, '127.0.0.1', r));

  console.log('Node.js - cancellation on client disconnect');
  console.log(
    `  upstream streams ${TOKENS} tokens x ${TOKEN_INTERVAL_MS}ms = ` +
      `${((TOKENS * TOKEN_INTERVAL_MS) / 1000).toFixed(1)}s of decode`,
  );
  console.log(`  client hangs up after ${CLIENT_HANGS_UP_AFTER_MS / 1000}s\n`);
  console.log(
    '  handler        upstream saw     tokens decoded  upstream ran   wasted',
  );
  console.log(`  ${'-'.repeat(70)}`);

  for (const path of ['/naive', '/cancelling']) {
    ledger.length = 0;
    await hangUpOn(path);
    const deadline = Date.now() + TOKENS * TOKEN_INTERVAL_MS + 1000;
    while (ledger.length === 0 && Date.now() < deadline) await sleep(50);
    const e = ledger[0] || { aborted: false, tokens: -1, seconds: NaN };
    const wasted = Math.max(0, e.seconds - CLIENT_HANGS_UP_AFTER_MS / 1000);
    console.log(
      `  ${path.padEnd(14)} ${(e.aborted ? 'cancelled' : 'nothing').padEnd(16)} ` +
        `${String(e.tokens).padStart(14)} ${e.seconds.toFixed(2).padStart(12)}s ` +
        `${wasted.toFixed(2).padStart(7)}s`,
    );
  }

  console.log();
  console.log("  'wasted' is decode time spent on a response nobody read. On a");
  console.log('  loaded server those KV blocks stayed allocated the whole time,');
  console.log('  so the scheduler could not admit somebody who was still waiting.');

  upstream.close();
  gateway.close();
}

main();
